compiler-core/src/type_/error.rs

```diff
@@ -697,7 +698,22 @@ pub enum Warning {
 }
 
 #[derive(Debug, Eq, PartialEq, Clone, Copy)]
+pub enum VariableKind {
```

Comment:
This doesn't seem right. Punning is not a kind of variable, rather it's property of the container a level above. Can this information be stored there instead?

Reply:
So this is needed specifically to track a variable's origin and report better error messages (maybe `VariableOrigin` would be a better name) on how to deal with unused variables coming from pattern matching on record constructors:
```rs
hint: match kind {
  VariableKind::Punned => "You can ignore it with an underscore: `{name}: _`.",
  VariableKind::Regular => "You can ignore it with an underscore: `_{name}`.",
}
```
Maybe I'm misunderstanding your question but I don't see how I could move it elsewhere since the warning needs this kind of information to display different hints

Reply:
Yeah I understand the need, but it's not correct as it's both misnamed and it's stored on the wrong AST node. Data about the layer above needs to be in the layer above.